src/cliffordsynthesis/encoding/MultiGateEncoder.cpp

```diff
@@ -132,19 +134,62 @@ void MultiGateEncoder::assertTwoQubitGateOrderConstraints(
   }
 
   // gate variables of the current and the next time step
-  const auto& current = vars.gC[pos][ctrl][trgt];
-  const auto& gSNow   = vars.gS[pos];
-  const auto& gCNext  = vars.gC[pos + 1];
+  const auto& gSNow  = vars.gS[pos];
+  const auto& gSNext = vars.gS[pos + 1];
+  const auto& gCNext = vars.gC[pos + 1];
 
   // two identical CNOTs may not be applied in a row because they would cancel.
-  lb->assertFormula(LogicTerm::implies(current, !gCNext[ctrl][trgt]));
+  lb->assertFormula(
+      LogicTerm::implies(vars.gC[pos][ctrl][trgt], !gCNext[ctrl][trgt]));
+  lb->assertFormula(
+      LogicTerm::implies(vars.gC[pos][trgt][ctrl], !gCNext[trgt][ctrl]));
 
   // if no gate is applied to both qubits, no CNOT on them can be applied in the
   // next time step.
-  const auto noneIndex     = gateToIndex(qc::OpType::None);
-  const auto noGate        = gSNow[noneIndex][ctrl] && gSNow[noneIndex][trgt];
+  // no gate on both qubits => no CNOT on them in the next time step.
+  constexpr auto noneIndex = gateToIndex(qc::OpType::None);
+  const auto     noGate    = gSNow[noneIndex][ctrl] && gSNow[noneIndex][trgt];
+  // similarly,
+  // H - X   c - H
+  //     | = |
+  // H - c   X - H
+  // i.e., Hadamards on both qubits => no CNOT on them in the next time step.
+  constexpr auto hIndex = gateToIndex(qc::OpType::H);
+  const auto     hh     = gSNow[hIndex][ctrl] && gSNow[hIndex][trgt];
+
   const auto noFurtherCnot = !gCNext[ctrl][trgt] && !gCNext[trgt][ctrl];
-  lb->assertFormula(LogicTerm::implies(noGate, noFurtherCnot));
+  lb->assertFormula(LogicTerm::implies(noGate || hh, noFurtherCnot));
+
+  // No Zs or Xs on both qubits after CNOT
+  constexpr auto xIndex  = gateToIndex(qc::OpType::X);
+  const auto     xxAfter = gSNext[xIndex][ctrl] && gSNext[xIndex][trgt];
+  constexpr auto zIndex  = gateToIndex(qc::OpType::Z);
+  const auto     zzAfter = gSNext[zIndex][ctrl] && gSNext[zIndex][trgt];
+  const auto cnotBefore  = vars.gC[pos][ctrl][trgt] || vars.gC[pos][trgt][ctrl];
+  lb->assertFormula(LogicTerm::implies(cnotBefore, !zzAfter && !xxAfter));
+
+  // No Zs or Xs on both qubits before CNOT
+  const auto xxBefore = gSNow[xIndex][ctrl] && gSNow[xIndex][trgt];
+  const auto zzBefore = gSNow[zIndex][ctrl] && gSNow[zIndex][trgt];
+  const auto cnotAfter =
+      vars.gC[pos + 1][ctrl][trgt] || vars.gC[pos + 1][trgt][ctrl];
+  lb->assertFormula(LogicTerm::implies(cnotAfter, !zzBefore && !xxBefore));
+}
```

Comment:
Maybe I just don't quite get it, but aren't these kind of incomplete and inefficient?
Doesn't some variation of this rule hold for any combination of Paulis before or after a CNOT? At the very least there should also be a rule for double Y gates. But, as said, I believe the rule can be more general than that.
Furthermore, the number of constraints could be reduced by combining the before and after rules (plus some special handling for the corner cases).